core/instance/GridBasedInterestManagement.js

```diff
@@ -0,0 +1,280 @@
+import EDictionary from "../../external/EDictionary";
+
+class GridCell {
+    constructor() {
+        this.numberOfStaticEntities = 0;
+        this.numberOfDynamicEntities = 0;
+        this.staticEntities = [];
+        this.dynamicEntities = [];
+        this.events = [];
+    }
+}
+
+Number.prototype.clamp = function(min, max) {
+  return Math.min(Math.max(this, min), max);
+};
+
+export default class GridBasedInterestManagement {
+    constructor(ID_PROPERTY_NAME, worldWidth, worldHeight, cellSize) {
+        this.ID_PROPERTY_NAME = ID_PROPERTY_NAME || "id";
+        this.events = new EDictionary(ID_PROPERTY_NAME);
+        this.worldWidth = worldWidth;
+        this.worldHeight = worldHeight;
+        this.gridCellSize = cellSize;
+        this.thresholdToBeConsideredABigEntity = 400
+        this.gridWidth = Math.floor(this.worldWidth / this.gridCellSize);
+        this.gridHeight = Math.floor(this.worldHeight / this.gridCellSize);
+        this.maxGridXIndex = this.gridWidth - 1;
+        this.maxGridYIndex = this.gridHeight - 1;
+        // this.AABBComparisonsThisTick = 0
+        // this.numberOfEntitiesNeedingToBeMovedToANewCellThisTick = 0
+
+        this.numberOfEntitiesInMasterList = 0;
+        this.masterEntityList = [];
+
+        this.numberOfBigEntities = 0;
+        this.bigEntityList = [];
+
+        this.cells = [];
+
+        for (let i = 0; i < this.gridWidth * this.gridHeight; i++) {
+            this.cells[i] = new GridCell();
+        }
+    }
+
+    create(ID_PROPERTY_NAME) {
+        return new GridBasedInterestManagement(ID_PROPERTY_NAME);
+    }
+
+    insertEntity(entity) {
+        const indexInMasterList = this.masterEntityList.push(entity) - 1;
+
+        this.numberOfEntitiesInMasterList += 1;
+
+        entity.nengiInternalMasterListIndex = indexInMasterList;
+
+        if (entity.visibilityRadius !== undefined && entity.visibilityRadius >= this.thresholdToBeConsideredABigEntity) {
+
+            const indexInBigEntityList = this.bigEntityList.push(entity) - 1;
+
+            this.numberOfBigEntities += 1;
+
+            entity.nengiinternalBigEntityListIndex = indexInBigEntityList;
+        } else {
+            const cellToBeAddedTo = this.getCellFromWorldPos(entity.x, entity.y)
+
+            entity.nengiInternalSpatialGridCell = cellToBeAddedTo
+
+            let indexInGridCellEntityList
+
+            if (entity.isStaticEntity) {
+                indexInGridCellEntityList = cellToBeAddedTo.staticEntities.push(entity) - 1
+                cellToBeAddedTo.numberOfStaticEntities += 1;
```

Comment:
would this always be equal to `cellsToBeAddedTo.staticEntities.length`? Is this an opt?

Reply:
yes this is an opt, "caching" the current length of all of these lists every time the length changes